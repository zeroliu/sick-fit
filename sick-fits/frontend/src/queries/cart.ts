import { MutationHookOptions, useMutation } from '@apollo/react-hooks';
import gql from 'graphql-tag';

import {
  Mutation,
  MutationAddToCartArgs,
  MutationRemoveFromCartArgs,
} from 'src/generated/graphql';

export const ADD_TO_CART_MUTATION = gql`
  mutation($data: AddToCartInput!) {
    addToCart(data: $data)
  }
`;
interface AddToCartMutationData {
  addToCart: Mutation['addToCart'];
}
export function useAddToCartMutation(
  options?: MutationHookOptions<AddToCartMutationData, MutationAddToCartArgs>,
) {
  return useMutation(ADD_TO_CART_MUTATION, options);
}

export const REMOVE_FROM_CART_MUTATION = gql`
  mutation($data: RemoveFromCartInput!) {
    removeFromCart(data: $data)
  }
`;
interface RemoveFromCartMutationData {
  removeFromCart: Mutation['removeFromCart'];
}
export function useRemoveFromCartMutation(
  options?: MutationHookOptions<
    RemoveFromCartMutationData,
    MutationRemoveFromCartArgs
  >,
) {
  return useMutation(REMOVE_FROM_CART_MUTATION, options);
}
