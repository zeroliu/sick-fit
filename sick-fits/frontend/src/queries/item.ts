import {
  useQuery,
  QueryHookOptions,
  MutationHookOptions,
  useMutation,
} from '@apollo/react-hooks';
import gql from 'graphql-tag';

import {
  Mutation,
  Query,
  QueryItemsArgs,
  QueryItemArgs,
  MutationCreateItemArgs,
  MutationUpdateItemArgs,
  MutationDeleteItemArgs,
} from 'src/generated/graphql';

export interface Item {
  id: string;
  title: string;
  price: number;
  description: string;
  image?: string;
  largeImage?: string;
}

const ALL_ITEMS_QUERY = gql`
  query($skip: Int, $take: Int) {
    items(skip: $skip, take: $take) {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;
interface AllItemsQueryData {
  items: Item[];
}
export function useAllItemsQuery(
  options?: QueryHookOptions<AllItemsQueryData, QueryItemsArgs>,
) {
  return useQuery(ALL_ITEMS_QUERY, options);
}

const ITEM_QUERY = gql`
  query($id: ID!) {
    item(id: $id) {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;
interface ItemQueryData {
  item?: Item;
}
export function useItemQuery(
  options?: QueryHookOptions<ItemQueryData, QueryItemArgs>,
) {
  return useQuery(ITEM_QUERY, options);
}

const ITEMS_CONNECTION_QUERY = gql`
  query {
    itemsConnection {
      totalCount
    }
  }
`;
interface ItemsConnectionQueryData {
  itemsConnection: Query['itemsConnection'];
}
export function useItemsConnectionQuery() {
  return useQuery<ItemsConnectionQueryData>(ITEMS_CONNECTION_QUERY);
}

const CREATE_ITEM_MUTATION = gql`
  mutation($data: CreateItemInput!) {
    createItem(data: $data) {
      id
    }
  }
`;
interface CreateItemMutationData {
  createItem: { id: number };
}
export function useCreateItemMutation(
  options?: MutationHookOptions<CreateItemMutationData, MutationCreateItemArgs>,
) {
  return useMutation(CREATE_ITEM_MUTATION, options);
}

const UPDATE_ITEM_MUTATION = gql`
  mutation($id: ID!, $data: UpdateItemInput!) {
    updateItem(id: $id, data: $data)
  }
`;
interface UpdateItemMutationData {
  updateItem: Mutation['updateItem'];
}
export function useUpdateItemMutation(
  options?: MutationHookOptions<UpdateItemMutationData, MutationUpdateItemArgs>,
) {
  return useMutation(UPDATE_ITEM_MUTATION, options);
}

const DELETE_ITEM_MUTATION = gql`
  mutation($id: ID!) {
    deleteItem(id: $id)
  }
`;
interface DeleteItemMutationData {
  deleteItem: Mutation['deleteItem'];
}
export function useDeleteItemMutation(
  options?: MutationHookOptions<DeleteItemMutationData, MutationDeleteItemArgs>,
) {
  return useMutation(DELETE_ITEM_MUTATION, options);
}
