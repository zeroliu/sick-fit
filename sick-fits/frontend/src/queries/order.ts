import { QueryHookOptions, useMutation, useQuery } from '@apollo/react-hooks';
import gql from 'graphql-tag';

import {
  MutationPayArgs,
  OrderItem as GeneratedOrderItem,
  Order as GeneratedOrder,
  User,
  QueryOrderArgs,
} from 'src/generated/graphql';

const PAY_MUTATION = gql`
  mutation($data: PayInput!) {
    pay(data: $data) {
      id
    }
  }
`;
interface PayMutationData {
  pay: Pick<GeneratedOrder, 'id'>;
}
export function usePayMutation(
  options?: QueryHookOptions<PayMutationData, MutationPayArgs>,
) {
  return useMutation(PAY_MUTATION, options);
}

const ORDER_QUERY = gql`
  query($orderId: ID!) {
    order(orderId: $orderId) {
      id
      paymentIntent
      total
      createdAt
      user {
        id
      }
      items {
        id
        title
        description
        price
        image
        quantity
      }
    }
  }
`;
type OrderItem = Pick<
  GeneratedOrderItem,
  'id' | 'title' | 'description' | 'price' | 'image' | 'quantity'
>;
export type Order = Pick<
  GeneratedOrder,
  'id' | 'paymentIntent' | 'total' | 'createdAt'
> & {
  user: Pick<User, 'id'>;
  items: OrderItem[];
};
interface OrderQueryData {
  order: Order;
}
export function useOrderQuery(
  options?: QueryHookOptions<OrderQueryData, QueryOrderArgs>,
) {
  return useQuery(ORDER_QUERY, options);
}
