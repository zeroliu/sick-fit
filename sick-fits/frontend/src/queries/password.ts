import gql from 'graphql-tag';
import { MutationHookOptions, useMutation } from '@apollo/react-hooks';

import {
  MutationResetPasswordArgs,
  MutationRequestResetArgs,
  User,
} from 'src/generated/graphql';

const REQUEST_RESET_MUTATION = gql`
  mutation($data: RequestResetInput!) {
    requestReset(data: $data)
  }
`;
interface RequestResetMutationData {
  requestReset: boolean;
}
export function useRequestResetMutation(
  options?: MutationHookOptions<
    RequestResetMutationData,
    MutationRequestResetArgs
  >,
) {
  return useMutation(REQUEST_RESET_MUTATION, options);
}

const RESET_PASSWORD_MUTATION = gql`
  mutation($data: ResetPasswordInput!) {
    resetPassword(data: $data) {
      id
    }
  }
`;
interface ResetPasswordMutationData {
  resetPassword: Pick<User, 'id'>;
}
export function useResetPasswordMutation(
  options?: MutationHookOptions<
    ResetPasswordMutationData,
    MutationResetPasswordArgs
  >,
) {
  return useMutation(RESET_PASSWORD_MUTATION, options);
}
