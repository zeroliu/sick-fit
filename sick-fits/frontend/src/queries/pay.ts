import { QueryHookOptions, useMutation } from '@apollo/react-hooks';
import gql from 'graphql-tag';

import { Mutation, MutationPayArgs } from 'src/generated/graphql';

const PAY_MUTATION = gql`
  mutation($data: PayInput!) {
    pay(data: $data) {
      id
    }
  }
`;
interface PayMutationData {
  pay: Mutation['pay'];
}
export function usePayMutation(
  options?: QueryHookOptions<PayMutationData, MutationPayArgs>,
) {
  return useMutation(PAY_MUTATION, options);
}
