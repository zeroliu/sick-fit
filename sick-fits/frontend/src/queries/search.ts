import gql from 'graphql-tag';
import { Item } from 'src/generated/graphql';
import { QueryOptions, ApolloQueryResult } from 'apollo-boost';
import { useApolloClient } from '@apollo/react-hooks';
import { useState } from 'react';

const SEARCH_ITEMS_QUERY = gql`
  query($searchTerm: String!) {
    items(searchTerm: $searchTerm) {
      id
      title
      image
    }
  }
`;
export type SearchItem = Pick<Item, 'id' | 'title' | 'image'>;
interface SearchItemsQueryData {
  items: SearchItem[];
}
interface SearchItemsQueryArgs {
  searchTerm: string;
}
type SearchItemsQueryOptions = QueryOptions<SearchItemsQueryArgs>;
export function useSearchItemsQueryBuilder(
  options?: SearchItemsQueryOptions,
): [
  (
    options?: Partial<SearchItemsQueryOptions>,
  ) => Promise<ApolloQueryResult<SearchItemsQueryData>>,
  { loading: boolean },
] {
  const client = useApolloClient();
  const [loading, setLoading] = useState(false);
  return [
    async (overriddenOptions) => {
      setLoading(true);
      const result = await client.query<
        SearchItemsQueryData,
        SearchItemsQueryArgs
      >({ query: SEARCH_ITEMS_QUERY, ...options, ...overriddenOptions });
      setLoading(false);
      return result;
    },
    { loading },
  ];
}
