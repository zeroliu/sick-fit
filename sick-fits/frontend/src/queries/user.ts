import gql from 'graphql-tag';
import {
  useQuery,
  useMutation,
  MutationHookOptions,
} from '@apollo/react-hooks';
import {
  MutationRegisterArgs,
  MutationSignInArgs,
  UserPermission,
  Mutation,
  MutationUpdatePermissionsArgs,
} from 'src/generated/graphql';

export interface User {
  id: string;
  email: string;
  name: string;
}

export interface UserWithPermissions extends User {
  permissions: UserPermission[];
}

const REGISTER_MUTATION = gql`
  mutation($data: RegisterInput!) {
    register(data: $data) {
      id
      name
      email
    }
  }
`;
interface RegisterMutationData {
  register: User;
}
export function useRegisterMutation(
  options?: MutationHookOptions<RegisterMutationData, MutationRegisterArgs>,
) {
  return useMutation(REGISTER_MUTATION, options);
}

const SIGN_IN_MUTATION = gql`
  mutation($data: SignInInput!) {
    signIn(data: $data) {
      id
      name
      email
    }
  }
`;
interface SignInMutationData {
  signIn: User;
}
export function useSignInMutation(
  options?: MutationHookOptions<SignInMutationData, MutationSignInArgs>,
) {
  return useMutation(SIGN_IN_MUTATION, options);
}

const SIGN_OUT_MUTATION = gql`
  mutation {
    signOut
  }
`;
interface SignOutMutationData {
  signOut: boolean;
}
export function useSignOutMutation(
  options?: MutationHookOptions<SignOutMutationData>,
) {
  return useMutation(SIGN_OUT_MUTATION, options);
}

export const ME_QUERY = gql`
  query {
    me {
      id
      name
      email
    }
  }
`;
interface MeQueryData {
  me: User;
}
export function useMeQuery() {
  return useQuery<MeQueryData>(ME_QUERY);
}

const USERS_QUERY = gql`
  query {
    users {
      id
      name
      email
      permissions
    }
  }
`;
interface UsersQueryData {
  users: UserWithPermissions[];
}
export function useUsersQuery() {
  return useQuery<UsersQueryData>(USERS_QUERY);
}

const UPDATE_PERMISSIONS_MUTATION = gql`
  mutation($data: UpdatePermissionsInput!) {
    updatePermissions(data: $data)
  }
`;
interface UpdatePermissionsMutationData {
  updatePermissions: Mutation['updatePermissions'];
}
export function useUpdatePermissionsMutation(
  options: MutationHookOptions<
    UpdatePermissionsMutationData,
    MutationUpdatePermissionsArgs
  >,
) {
  return useMutation(UPDATE_PERMISSIONS_MUTATION, options);
}
