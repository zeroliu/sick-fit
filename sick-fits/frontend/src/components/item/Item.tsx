import React from 'react';
import { Item as ItemType } from 'src/queries/item';
import { Item as StyledItem } from './item_styles';
import { Title } from 'src/components/styles/Title';
import Link from 'next/link';
import { PriceTag } from 'src/components/styles/PriceTag';
import { formatMoney } from 'src/lib/format_money';
import { DeleteItem } from '../delete_item/DeleteItem';

interface Props {
  data: ItemType;
}

export const Item: React.FC<Props> = ({ data }) => {
  return (
    <StyledItem>
      {data.image && <img src={data.image} alt={data.title}></img>}
      <Title>
        <Link
          href={{
            pathname: '/item',
            query: { id: data.id },
          }}>
          <a>{data.title}</a>
        </Link>
      </Title>
      <PriceTag>{formatMoney(data.price)}</PriceTag>
      <p>{data.description}</p>
      <div className='buttonList'>
        <Link
          href={{
            pathname: '/update',
            query: { id: data.id },
          }}>
          <a>Edit ✏️</a>
        </Link>
        <button>Add to Cart</button>
        <DeleteItem id={data.id}>Delete</DeleteItem>
      </div>
    </StyledItem>
  );
};
