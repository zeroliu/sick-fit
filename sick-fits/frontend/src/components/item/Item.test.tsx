import { MockedProvider } from '@apollo/react-testing';
import { render } from '@testing-library/react';
import React from 'react';

import { Item } from './Item';
import { Item as ItemType } from 'src/queries/item';

const fakeItem: ItemType = {
  id: '1',
  title: 'fake item',
  description: 'this is a fake item',
  price: 1234,
  image: 'foo.jpg',
  largeImage: 'large_foo.jpg',
};

describe('<Item />', () => {
  it('matches the snapshot', () => {
    const { container } = render(
      <MockedProvider>
        <Item data={fakeItem} />
      </MockedProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});
