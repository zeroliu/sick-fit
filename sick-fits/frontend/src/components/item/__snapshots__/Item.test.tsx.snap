// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Item /> matches the snapshot 1`] = `
<div>
  <div
    class="item_styles__Item-sc-1wrkfgx-0 aVlSA"
  >
    <img
      alt="fake item"
      src="foo.jpg"
    />
    <h3
      class="Title-sc-1002kw6-0 eBuTHz"
    >
      <a
        href="/item?id=1"
      >
        fake item
      </a>
    </h3>
    <span
      class="PriceTag-sc-145cui2-0 cUpSNy"
    >
      $12.34
    </span>
    <p>
      this is a fake item
    </p>
    <div
      class="buttonList"
    >
      <a
        href="/update?id=1"
      >
        Edit ✏️
      </a>
      <button>
        Add To Cart
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
</div>
`;
